### PR TITLE
Starter Packs

### PR DESCRIPTION
**Pull request by haileyok**
### Other PRs

This PR relies on two other PRs (which have been merged into this PR). Don't merge this in until we look at both of those and/or revert them being merged into this PR.

https://github.com/bluesky-social/social-app/pull/4403
https://github.com/bluesky-social/social-app/pull/4533

### Wizard

- [x] Details
- [x] Profiles
  - [x] Search
  - [x] List of those that have been added
  - [x] Use `typeahead` query over follows **Pending**
- [x] Feeds
- [x] Edit existing starter packs

### Logged In

- [x] Starter pack list on profiles
- [x] QR code generation for sharing starter packs
  - [x] Tweak style of this thing
  - [x] Dark mode
- [x] Starter pack screen showing name, description, profiles, and feeds
- [x] Sharing by a short link
- [x] Newskie badge on profiles
  - [x] Dialog that shows the starter pack used if applicable

### Logged Out

- [x] Landing
  - [x] iOS (App Clip) **Pending**
  - [x] Android (Android Referrer API)
  - [x] Web (Basic)
- [x] Follow suggested profiles on account creation
- [x] Pin suggested feeds
- [x] Set created at date
- [x] Land on the first-most feed if there were feeds in the pack
- [x] Land on the following feed if there were only profiles and there were 25 or more profiles in the pack

### Stats

- [x] Created an account w/ a starter pack
  - [x] Which one?
- [x] Followed accounts or pinned feeds from the starter pack page

### What the fucks

- [x] Make sure nothing weird happens if the starter pack still exists but the list is deleted
- [x] Make sure we only fetch 50 profiles even if the list has some other number (and restrict the follows to that limit)

### CODE REVIEW

**Comment by haileyok** [comment]
Lets start reviews here now. Going to continue iterating in separate PRs.

Note you need to do a `yarn` because this relies on an RC atproto package

File: eas.json
Comment (haileyok) [inline eas.json:66]: dont forget to revert

File: src/screens/Profile/Header/DisplayName.tsx
Comment (estrattonbailey) [inline src/screens/Profile/Header/DisplayName.tsx:1]: Nit: could bump this from the PR

File: src/screens/Profile/Header/Handle.tsx
Comment (estrattonbailey) [inline src/screens/Profile/Header/Handle.tsx:1]: Honestly we could extract this too and merge it prior to starter packs, right?

**Review by estrattonbailey — COMMENTED** [review]
Gargantuan task, great work — I'll keep reviewing in a bit

File: bskyweb/cmd/bskyweb/server.go
Comment (haileyok) [reply bskyweb/cmd/bskyweb/server.go:225]: So for signed in users we use `starter-pack` and for signed out ones its `start`. That's because the App Clip is going to try and present itself on a specific URL path.

File: modules/expo-bluesky-swiss-army/ios/ExpoBlueskySwissArmy.podspec
Comment (haileyok) [reply modules/expo-bluesky-swiss-army/ios/ExpoBlueskySwissArmy.podspec:5]: Yea, we keep adding little modules for small things. Like ideally we can just throw this Google Play referrer stuff into this toolkit. I'm just holding off because there's a whole PR for referrers https://github.com/bluesky-social/social-app/pull/4514 that I'll just put into this module once we merge. Trying to simplify that down a bit.

File: src/components/hooks/useStarterPackEntry.native.ts
Comment (haileyok) [reply src/components/hooks/useStarterPackEntry.native.ts:43]: Let's use a ref. Don't want to update the state immediately because - primarily on Android - the request for referrer is async, and we don't want the splash screen to disappear until we get this value. Good callout.

File: src/lib/strings/time.ts
Comment (haileyok) [reply src/lib/strings/time.ts:1]: Yea I have that on a little todo list. There's another `ago` PR though that is kind of waiting to get merged in - our `ago` test right now is actually disabled.

File: src/routes.ts
Comment (haileyok) [reply src/routes.ts:4]: If you give an array to this, it will match multiple URL paths to the same React route. So in this case both `https://bsky.app/` and `https://bsky.app/profile/start/:name/:rkey` will route to `Home`. That we handle here

https://github.com/bluesky-social/social-app/blob/077da0830924d4c040ef095d8349f13fdfdf4372/src/lib/routes/router.ts#L3

File: src/screens/Profile/Header/Handle.tsx
Comment (haileyok) [reply src/screens/Profile/Header/Handle.tsx:1]: Will need to ask Devin, but yea I think so. I noticed while I was testing that someone else's new account had a little badge. Think it gets handled on the backend.

File: app.config.js
Comment (haileyok) [inline app.config.js:49]: note to self: let's rm these before a merge

File: app.config.js
Comment (haileyok) [inline app.config.js:62]: another note: we need to add this back

File: src/components/ProfileCard.tsx
Comment (haileyok) [reply src/components/ProfileCard.tsx:32]: Ahh yep. Good callout on the moderationOpts, forgot we'd need to do that in this component.

File: src/components/hooks/useStarterPackEntry.native.ts
Comment (haileyok) [reply src/components/hooks/useStarterPackEntry.native.ts:40]: This is only for Android. We can't clear the referrer info once it gets set, remains for 90 days after installing the app. So we just keep track of it here to not re-trigger the same starter pack.

File: src/components/hooks/useStarterPackEntry.native.ts
Comment (haileyok) [reply src/components/hooks/useStarterPackEntry.native.ts]: Great callout! Added tests for this as well as others.

File: src/components/LinearGradientBackground.tsx
Comment (haileyok) [reply src/components/LinearGradientBackground.tsx:19]: Oh, I like `sky`. I change it but let me know if there's an easier way to use it?

File: src/view/screens/Home.tsx
Comment (haileyok) [inline src/view/screens/Home.tsx:129]: note that this is a hack and needs work. it doens't actually work right now, though it worked on the local dev env. not sure why, maybe something with how we add discover?

File: src/view/com/util/List.web.tsx
Comment (haileyok) [inline src/view/com/util/List.web.tsx:76]: All the changes in this file are in a separate PR for @gaearon to review https://github.com/bluesky-social/social-app/pull/4403

File: src/lib/generate-starterpack.ts
Comment (haileyok) [reply src/lib/generate-starterpack.ts:69]: nice

File: src/lib/statsig/events.ts
Comment (haileyok) [reply src/lib/statsig/events.ts:61]: Added to maybe make it easier to sort/view the data. Would need to visit the URI whereas if we pass it in the event we can immediately get some context within Statsig 🤷

File: src/screens/Profile/Header/Handle.tsx
Comment (haileyok) [reply src/screens/Profile/Header/Handle.tsx:1]: Confirmed we're good to do this!

File: src/components/hooks/useStarterPackEntry.native.ts
Comment (haileyok) [reply src/components/hooks/useStarterPackEntry.native.ts:33]: Thinking about another way to handle this thats a little clearer.

File: src/lib/generate-starterpack.ts
Comment (haileyok) [reply src/lib/generate-starterpack.ts:51]: The only reason I didn't do that was because it felt a little weird to need a hook  for this in particular. Not opposed to putting it in a mutation though.

File: src/components/hooks/useStarterPackEntry.native.ts
Comment (gaearon) [inline src/components/hooks/useStarterPackEntry.native.ts:24]: Is there any risk of this failing? Or stalling for a long time?

File: src/components/hooks/useStarterPackEntry.native.ts
Comment (gaearon) [inline src/components/hooks/useStarterPackEntry.native.ts:30]: Same question here. If anything fails we'll be stuck with a splash screen so we need high confidence, right?

File: src/screens/Onboarding/StepFinished.tsx
Comment (gaearon) [inline src/screens/Onboarding/StepFinished.tsx:22]: Is this being saved per-device (prefs API) rather than per-account? Is it being synced between tabs?

File: src/view/screens/Home.tsx
Comment (gaearon) [inline src/view/screens/Home.tsx:59]: We should never have empty catches when there is _our_ code in the try — it'll swallow legit mistakes like reference/type errors.

File: src/App.native.tsx
Comment (gaearon) [inline src/App.native.tsx:102]: Flagging as a concern that we need to make sure `hasCheckedReferrer` _definitely_ resolves to `true` in reasonable time (no matter what errors etc) because otherwise we'll get stuck on the splash screen which would be Very Bad.

File: src/view/screens/Home.tsx
Comment (gaearon) [inline src/view/screens/Home.tsx:56]: What does this logic do? Which case is it for?

File: src/view/screens/Home.tsx
Comment (gaearon) [inline src/view/screens/Home.tsx:102]: Can we move this logic to `selected-feed` instead where we're determining the feed to start on?

**Review by gaearon — COMMENTED** [review]
flagging some things that i'm not sure about, will look deeper tomorrow. overall i'd like to better understand the proposed state management model — when do the "current" and "used" started packs get cleared out, what are the expectations wrt how it interacts with account switching, what are the expectations re: how it interacts with the app startup.

File: src/routes.ts
Comment (estrattonbailey) [reply src/routes.ts:4]: Nice thanks

File: src/lib/strings/starter-pack.ts
Comment (estrattonbailey) [inline src/lib/strings/starter-pack.ts:53]: Do we need to do any more validation of this data? Looks like we'd just navigate to a route that's not found, so maybe nbd

File: src/lib/strings/starter-pack.ts
Comment (estrattonbailey) [inline src/lib/strings/starter-pack.ts:18]: I think this won't work with handles that contain hyphens, right

File: src/view/screens/Home.tsx
Comment (estrattonbailey) [inline src/view/screens/Home.tsx:62]: Nit: `new` is explicitly defined in `shell/logged-out` as a special case, would be nice to have `starterpack` in there too

File: src/screens/Onboarding/StepFinished.tsx
Comment (estrattonbailey) [inline src/screens/Onboarding/StepFinished.tsx:99]: I think this should be Discover? That's the first pinned feed, or should be.

File: src/view/com/notifications/FeedItem.tsx
Comment (estrattonbailey) [inline src/view/com/notifications/FeedItem.tsx:101]: Just marking todo as a double check 👍

File: src/state/preferences/used-starter-packs.tsx
Comment (estrattonbailey) [inline src/state/preferences/used-starter-packs.tsx:1]: This is basically just for Android right? Just confirming I grok it

File: src/components/hooks/useStarterPackEntry.native.ts
Comment (estrattonbailey) [reply src/components/hooks/useStarterPackEntry.native.ts:40]: yeah +1 — 500ms timeout seems good, could pry use a try/catch and logging here too

File: src/components/hooks/useStarterPackEntry.native.ts
Comment (estrattonbailey) [inline src/components/hooks/useStarterPackEntry.native.ts:10]: Very nit: naming here kinda odd to me, like it's called "entry" and it sets "active pack" and returns "has checked referrer". It's like "use init starter pack from referrer" and returns "is done" or "ready". Very nit, just spitballing.

File: src/components/hooks/useStarterPackEntry.ts
Comment (estrattonbailey) [inline src/components/hooks/useStarterPackEntry.ts:20]: Any risk of a flash of a diff screen here since this happens in an effect, or is that handled by routing on web?

File: src/lib/generate-starterpack.ts
Comment (estrattonbailey) [inline src/lib/generate-starterpack.ts:114]: I learned something recently and forgot to share with the group, but Sentry filters out anything that looks like an `Error` class, even if it's been stringified (which we do).

For errors that won't contain sensitive PII, `logger.error(e, { context: 'what happened' })` is often better so that Sentry can do its thing with the Error class and we still get detail on what's up.

Another option if you want a consistent failure message in Sentry — which we probably do here bc there are multiple async calls — is use `safeMessage` breadcrumb field. That one is allow-listed in Sentry and won't be sanitjized.

File: src/lib/generate-starterpack.ts
Comment (estrattonbailey) [inline src/lib/generate-starterpack.ts:92]: Is this translatable OK?

File: src/screens/Login/LoginForm.tsx
Comment (estrattonbailey) [inline src/screens/Login/LoginForm.tsx:121]: Why is this here? Can we set this in the "entry" hook?

File: src/screens/Onboarding/StepFinished.tsx
Comment (estrattonbailey) [inline src/screens/Onboarding/StepFinished.tsx:75]: Makes sense to do this up front, and we haven't done much error handling here before, but what do you think about more gracefully handling failures of this data separately?

Like if we get this data up front, _then_ handle it later on. But if this data fails, existing code continues to run as it does today.

File: src/lib/generate-starterpack.ts
Comment (haileyok) [reply src/lib/generate-starterpack.ts:114]: Nice, fixed these in the other PR. There were two that I remember including this one, keep an eye out though for others and I'll go through as well.

File: src/screens/Onboarding/StepFinished.tsx
Comment (haileyok) [reply src/screens/Onboarding/StepFinished.tsx]: Totally wiped out storing anything here, just using a context higher up the tree.

File: src/components/hooks/useStarterPackEntry.native.ts
Comment (haileyok) [reply src/components/hooks/useStarterPackEntry.native.ts:40]: yup, i made it 500ms, though sould never last that long.

idt there is anything we'll catch with a try/catch here. `NSUserDefaults` will just either give us a string or null - either of which is a valid response here for us since we don't require the string to be set.

File: src/screens/Onboarding/StepFinished.tsx
Comment (haileyok) [reply src/screens/Onboarding/StepFinished.tsx:75]: not opposed. best probably to get the user through signup successfully. i can think of a case where a deletion might happen before the user even gets through signup - and if that happens then they're screwed until they decide to not use the starter pack 😅  good call

Comment (haileyok) [reply src/screens/Onboarding/StepFinished.tsx:75]: [`e9dcb8d` (#4548)](https://github.com/bluesky-social/social-app/pull/4548/commits/e9dcb8d5ed73a3897a7e2105bf6d688995bc1b8e)

File: src/components/StarterPack/NewskieDialog.tsx
Comment (haileyok) [inline src/components/StarterPack/NewskieDialog.tsx:1]: this got duped when i merged the other PR. Removed it in the next PR.

File: src/view/shell/desktop/LeftNav.tsx
Comment (haileyok) [inline src/view/shell/desktop/LeftNav.tsx:114]: I'm kind of surprised we've never ran into this case before. We could maybe create a context that lets us toggle this button's visibility?

File: package.json
Comment (gaearon) [inline package.json:52]: Planning to cut a patch or leave as is?

Comment (haileyok) [reply package.json:52]: Likely can cut beforehand, though this RC was also just cut today so it should be fine.

Anyway I think the PR is in a good state but will ping @devinivy to make sure. Paul should know how to do that. https://github.com/bluesky-social/atproto/pull/2553

File: src/screens/Onboarding/StepFinished.tsx
Comment (gaearon) [inline src/screens/Onboarding/StepFinished.tsx]: Need to change this logic to always do  this.

File: src/screens/StarterPack/Wizard/State.tsx
Comment (gaearon) [inline src/screens/StarterPack/Wizard/State.tsx:108]: This potentially mutates the previous value due to https://github.com/bluesky-social/social-app/pull/4332/files#diff-1790dc6c929b51e7ef5ae32e12e420f7a35090cef3e054844e137e6e9cc34fcaR50. If this is intended to be recalculated regardless of the action, seems like you'd always want to clone `...state` in the beginning.

File: src/screens/StarterPack/Wizard/State.tsx
Comment (gaearon) [inline src/screens/StarterPack/Wizard/State.tsx:152]: `createInitialState()` here will run on every render. Instead you can do `React.useReducer(reducer, null, createInitialState)`

File: src/screens/StarterPack/Wizard/StepFeeds.tsx
Comment (gaearon) [inline src/screens/StarterPack/Wizard/StepFeeds.tsx:44]: Nit: i think you can `useThrottledValue` if you want, it should update as you type

File: src/state/queries/starter-packs.ts
Comment (gaearon) [inline src/state/queries/starter-packs.ts:56]: TODO?

File: src/state/queries/starter-packs.ts
Comment (gaearon) [inline src/state/queries/starter-packs.ts:109]: TODO?

File: src/state/queries/starter-packs.ts
Comment (gaearon) [inline src/state/queries/starter-packs.ts:218]: TODO

File: src/view/screens/Home.tsx
Comment (gaearon) [inline src/view/screens/Home.tsx:89]: Is this set of changes still needed? I think maybe not. If I'm right, let's fully undo this file.

**Review by gaearon — APPROVED** [review]
lgtm with some changes

File: src/screens/StarterPack/Wizard/State.tsx
Comment (haileyok) [reply src/screens/StarterPack/Wizard/State.tsx]: ah yea, not necessary. `canNext` is only used for disabling or enabling the `Next` button. Button can use a prop `disabled={!state.canNext || state.processing}` instead.

File: src/view/screens/Home.tsx
Comment (haileyok) [reply src/view/screens/Home.tsx:89]: Oh yea, should be reverted 👍

**Pull request merged**
